text/3898-thin-pointers-with-inline-metadata.md

```diff
@@ -0,0 +1,370 @@
+/// constructed first, then coerced (unsized) to `Thin<U>`
+/// (aka `Thin<U, U>`)
+#[repr(C)]
+pub struct Thin<T: Pointee, U: Pointee = T> {
+    metadata: U::Metadata,
+    data: EraseMetadata<T>,
+}
+
+// The size is known via reading its metadata.
+impl<U: Pointee> ValueSized for Thin<U> {}
+
+// Value accesses
+impl<U: Pointee> ops::Deref for Thin<U> {
+    type Target = U;
+    fn deref(&self) -> &U;
+}
+impl<U: Pointee> ops::DerefMut for Thin<U> {
+    fn deref_mut(&mut self) -> &mut U;
+}
+```
+
+### `EraseMetadata<T>`
+`EraseMetadata<T>` is a wrapper of (maybe unsized) `T`, which ignores the metadata of `T`.
+
+For example, both `&EraseMetadata<dyn Trait>` and `&EraseMetadata<[u8]>` have the same size as a thin
+pointer `&()`.
+
+```rust
+/// A wrapper that ignores the metadata of a type.
+#[lang = "erase_metadata"]
+#[repr(transparent)]
+pub struct EraseMetadata<T: Pointee>(T);
+
+// For sized types, `EraseMetadata` is a simple wrapper.
+impl<T: Sized> Sized for EraseMetadata<T> {}
+impl<T: Sized> ops::Deref for EraseMetadata<T> {
+    type Target = T;
+    fn deref(&self) -> &T {
+        &self.0
+    }
+}
+impl<T: Sized> ops::DerefMut for EraseMetadata<T> {
+    fn deref_mut(&mut self) -> &mut T {
+        &mut self.0
+    }
+}
+
+impl<T: Sized> EraseMetadata<T> {
+    /// Wrap a sized value into an `EraseMetadata`.
+    pub fn new(inner: T) -> EraseMetadata<T> {
+        EraseMetadata(inner)
+    }
+    /// Unwrap a sized value from an `EraseMetadata`.
+    pub fn into_inner(self) -> T {
+        self.0
+    }
+}
+
+// For unsized types, `EraseMetadata` is completely opaque because it is unsafe
+// to read the inner value without the metadata.
+
+// The size is unknown because the metadata is erased.
+impl<T: MetaSized> PointeeSized for EraseMetadata<T> {}
+```
+
+## Value constructions
+For a sized type `Thin<T>`, it can be constructed with `Thin::<T>::new`.
+For an unsized (`MetaSized`) type `Thin<U>`, in general, it requires 3 steps to construct
+a `Thin<U>` on stack or on heap:
+- construct a sized value of `Thin<T, U>` via `Thin::<T, U>::new_unsized` (where `T: Unsize<U>`).
+- obtain a pointer (i.e., `&`, `&mut`, `Box`, `Rc`, `Arc`, etc.) of `Thin<T, U>` via
+  their constructors.
+- coerce the pointer of `Thin<T, U>` to the pointer of `Thin<U>`.
+
+Here are the APIs related to value constructions mentioned above:
+```rust
+impl<T: Sized> Thin<T> {
+    /// Create a sized `Thin<T>` value, which is a simple wrapper of `T`
+    pub fn new(value: T) -> Thin<T> {
+        Self {
+            metadata: (), // Sized type `T` has empty metadata
+            data: EraseMetadata(value),
+        }
+    }
+}
+
+impl<T: Sized, U: Pointee> Thin<T, U> {
+    /// Create a sized `Thin<T, U>` value with metadata of unsized type `U`,
+    /// which can be coerced (unsized) to `Thin<U>`
+    pub fn new_unsized(value: T) -> Self
+    where
+        T: Unsize<U>,
+    {
+        Self {
+            metadata: ptr::metadata(&value as &U),
```

Comment:
I think this puts constraints on what metadata can be, `ptr::metadata(&value as &U)` has to always return the same value, so it has to be pure, not depend on address etc. This is plausibly fine but definitely needs to be documented.